File: src/python/exp_defs.py
#!/usr/bin/env python
# -*- coding: utf-8; tab-width: 4; -*-
# @(#) exp_defs.py  Time-stamp: <Julian Qian 2015-12-18 16:39:54>
# Copyright 2015 Julian Qian
# Author: Julian Qian <junist@gmail.com>
# Version: $Id: exp_defs.py,v 0.1 2015-11-13 17:09:51 jqian Exp $
#

import collections
import json

from exp_log import logger


class BucketError(Exception):
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return repr(self.value)


class Bucket(object):
    def __init__(self, offset, assign_id):
        self.offset = offset
        self.assign_id = assign_id


class Buckets(object):
    def __init__(self, bucketRanges, init_id=-1):
        self.buckets = []
        self.init_id = init_id
        for br in bucketRanges:
            for i in range(br.start, br.end):
                self.buckets.append(Bucket(i, self.init_id))

    def __str__(self):
        return '<%d buckets, inited %d>' % (len(self.buckets), self.init_id)

    def get_bucket_ranges(self):
        '''@return a list of BucketRange
        '''
        # TODO O(1) or O(n)?
        bucketRanges = []
        for bucket in self.buckets:
            bucketRanges.append(BucketRange(bucket.offset, bucket.offset+1))
        return self._merge_bucket_ranges(bucketRanges)

    def buckets_num(self):
        return len(self.buckets)

    @classmethod
    def from_num(cls, start_offset, num, init_id=-1):
        bucketRanges = [BucketRange(start_offset, start_offset + num)]
        return cls(bucketRanges, init_id)

    def _merge_bucket_ranges(self, bucketRanges):
        '''merge bucket ranges

        @param BucketRange list
        @return merged BucketRange list
        '''
        ret_brs = []
        alen = len(bucketRanges)
        if alen > 1:
            for i in range(1, alen):
                if bucketRanges[i-1].end == bucketRanges[i].start: # merge
                    bucketRanges[i].start = bucketRanges[i-1].start
                else:
                    ret_brs.append(bucketRanges[i-1])
                if i == alen-1:
                    ret_brs.append(bucketRanges[i])
        else:
            ret_brs = bucketRanges
        return ret_brs

    def assign(self, assign_id, assign_num):
        '''assign buckets to domain/experiments

        @return a list of assigned BucketRanges
        '''
        brs = []
        for bucket in self.buckets:
            if bucket.assign_id == assign_id:
                assign_num -= 1
                brs.append(BucketRange(bucket.offset, bucket.offset+1))
            elif bucket.assign_id == self.init_id: # empty bucket, assign it
                bucket.assign_id = assign_id
                assign_num -= 1
                brs.append(BucketRange(bucket.offset, bucket.offset+1))
            if assign_num == 0:
                break
        if assign_num != 0:
            raise BucketError('remaining %d buckets' % assign_num)
        return self._merge_bucket_ranges(brs)

    def load(self, assign_id, bucketRanges):
        '''load buckets for domain/experiments

        @param buckets, a list of assigned buckets
        '''
        loaded_cnt = 0
        for br in bucketRanges:
            if br.start < self.buckets[0].offset or \
               br.end > self.buckets[-1].offset:
                # raise exception?
                logger.error('bucket range %s error, exceed ranges!', br)
                raise BucketError('bucket range %s error, exceed ranges!' % br)
            else:
                for bkt in self.buckets:
                    if br.start <= bkt.offset < br.end:
                        if bkt.assign_id == self.init_id:
                            bkt.assign_id = assign_id
                            # logger.debug('assign bucket %d to %d',
                            #              bkt.offset, assign_id)
                            loaded_cnt += 1
                        else:
                            logger.error('bucket %d has been assigned to %d',
                                         bkt.offset, bkt.assign_id)
                            raise BucketError('bucket %d has been assigned to %d' \
                                              % (bkt.offset, bkt.assign_id))
        return loaded_cnt


# BucketRange = collections.namedtuple('BucketRange', 'start,end')
class BucketRange(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end

    @classmethod
    def from_json(cls, json):
        return cls(json[0], json[1])

    def to_json(self):
        return [self.start, self.end]

    def __eq__(self, o):
        return self.start == o.start and self.end == o.end

    def __str__(self):
        return '<BucketRange [%d-%d)>' % (self.start, self.end)

    def get_buckets_num(self):
        return self.end - self.start


class Domain(object):
    def __init__(self, id, bucketRanges):
        self.id = id
        self.buckets = Buckets(bucketRanges, -10000 + self.id)

    @classmethod
    def from_num(cls, id, offset, buckets_num):
        bucketRanges = [BucketRange(offset, offset+buckets_num)]
        return cls(id, bucketRanges)

    def to_json(self):

        bucketRanges = self.buckets.get_bucket_ranges()
        if len(bucketRanges) > 1:
            logger.warn('domain %d, something wrong? to many buckets',
                        self.id)
        return {'id': self.id,
                'range': bucketRanges[0].to_json()}

    def assign(self, assign_id, buckets_num):
        return self.buckets.assign(assign_id, buckets_num)

    def __str__(self):
        return "<Domain %d: %s>" % (self.id, self.buckets)


class Layer(object):
    def __init__(self, id, domain):
        '''
        @param domain, Domain
        '''
        self.id = id
        self.domain_id = domain.id
        bucketRanges = domain.buckets.get_bucket_ranges()
        self.buckets = Buckets(bucketRanges, -1000 + self.id)

    def to_json(self):
        return {'id': self.id,
                'domain_id': self.domain_id}

    def load(self, assign_id, bucketRanges):
        return self.buckets.load(assign_id, bucketRanges)

    def assign(self, assign_id, buckets_num):
        return self.buckets.assign(assign_id, buckets_num)

    def __str__(self):
        return "<Layer %d (D%d)>" % (self.id, self.domain_id)


class Condition(object):
    def __init__(self, name, args):
        '''@param args, list of condition string
        '''
        self.name = name
        self.args = args

    @classmethod
    def from_json(cls, json):
        return cls(json['name'], json['args'])

    def to_json(self):
        return {'name': self.name,
                'args': self.args}

    def __str__(self):
        return "<Condition %s>" % self.name


class Parameter(object):
    def __init__(self, name, value, type=None):
        self.name = name
        if type:
            self.type = self._trans_type(type)
        else:
            self.type = self._valid_type(value)
        self.value = self._trans_value(value, self.type)

    @staticmethod
    def _valid_type(value):
        t = 'string'
        if isinstance(value, bool):
            t = 'bool'
        elif isinstance(value, int):
            t = 'int'
        elif isinstance(value, float):
            t = 'double'
        return t

    @staticmethod
    def _trans_type(t):
        if t in ('bool', 'int', 'string', 'double'):
            return str(t)
        else:
            return 'string'

    @staticmethod
    def _trans_value(value, type):
        if Parameter._valid_type(value) == type:
            return value
        # else value is a string
        ret = None
        if type == 'bool':
            if value.lower() == 'true':
                ret = True
            else:
                ret = False
        elif type == 'int':
            ret = int(value)
        elif type == 'double':
            ret = float(value)
        else:
            ret = str(value)
        return ret

    @classmethod
    def from_json(cls, json):
        return cls(json['name'], json['value'], json['type'])

    def to_json(self):
        return {'name': self.name,
                'value': self.value,
                'type': self.type}

    def __str__(self):
        return "<Parameter %s: %s>" % (self.name, self.value)


class Diversion(object):
    @staticmethod
    def valid(d):
        if d in ('uuid', 'user', 'random'):
            return d
        else:
            return 'random'


class Experiment(object):
    def __init__(self, id, layer_id, diversion,
                 start_time, end_time,
                 ranges, parameters, conditions):
        '''
        @diversion Diversion
        @ranges BucketRanges
        @parameters Parameters
        @conditions Conditions
        '''
        self.id = id
        self.layer_id = layer_id
        self.diversion = diversion
        self.start_time = start_time
        self.end_time = end_time
        self.ranges = ranges
        self.parameters = parameters
        self.conditions = conditions

    @classmethod
    def from_json(cls, json):
        ranges = [BucketRange.from_json(i) for i in json['ranges']]
        parameters = [Parameter.from_json(i) for i in json['parameters']]
        conditions = [Condition.from_json(i) for i in json['conditions']]
        return cls(json['id'], json['layer_id'], json['diversion'],
                   json['start_time'], json['end_time'],
                   ranges, parameters, conditions)

    @classmethod
    def from_db(cls, row):
        diversion = Diversion.valid(row.diversion)
        logger.debug('row: %s', row)
        ret = json.loads(row.parameters) if row.parameters else {}
        parameters = [Parameter(name, value) for name, value in ret.items()]
        ret = json.loads(row.conditions) if row.conditions else {}
        conditions = [Condition(name, args) for name, args in ret.items()]
        return cls(row.id, row.layer_id, diversion,
                   '%s' % row.start_time, '%s' % row.end_time,
                   None, parameters, conditions)

    def to_json(self):
        # return self.__dict__
        return {'id': self.id,
                'layer_id': self.layer_id,
                'diversion': self.diversion,
                'start_time': self.start_time,
                'end_time': self.end_time,
                'ranges': [i.to_json() for i in self.ranges],
                'parameters': [i.to_json() for i in self.parameters],
                'conditions': [i.to_json() for i in self.conditions]}

    def set_bucket_ranges(self, bucketRanges):
        '''fill BucketRanges after calling `from_db`
        '''
        self.ranges = bucketRanges

    def get_buckets_num(self):
        num = 0
        for br in self.ranges:
            num += br.get_buckets_num()
        return num

    def __eq__(self, o):
        if self.id == o.id and \
           self.layer_id == o.layer_id:
            if len(self.ranges) == len(o.ranges):
                for i, r in enumerate(self.ranges):
                    if o.ranges[i] != r:
                        return False
                return True
        return False

    def __str__(self):
        return "<Experiment %d, Layer %d, Buckets %d)>" % (
            self.id, self.layer_id, self.get_buckets_num())


def main():
    pass

if __name__ == "__main__":
    main()

File: src/php/exp_defs.php
<?php
// @(#) exp_defs.php  Time-stamp: <Julian Qian 2015-12-18 15:55:34>
// Copyright 2015 Julian Qian
// Author: Julian Qian <junist@gmail.com>
// Version: $Id: exp_defs.php,v 0.1 2015-12-02 16:12:19 jqian Exp $
//

if (!class_exists("Logger")) {
  require_once __DIR__ . "/exp_log.php";
}

define("BUCKETS_NUM_MAX", 10000);


class BucketRange {
  public
  function __construct($start=0, $end=0) {
    $this->start = intval($start);
    $this->end = intval($end);
  }

  public static
  function fromJson($json) {
    return new self($json[0], $json[1]);
  }

  public
  function num() {
    return $this->end - $this->start;
  }

  public
  function contain($idx) {
    return $idx >= $this->start && $idx < $this->end;
  }

  public
  function getStart() {
    return $this->start;
  }

  public
  function getEnd() {
    return $this->end;
  }

  public
  function toString() {
    return sprintf("<BucketRange %d~%d>", $this->start, $this->end);
  }
}

class Buckets {
  public
  function __construct($bucketRange) {
    $this->range = $bucketRange;
    $this->buckets = array();
    for ($i=0; $i<$bucketRange->num(); $i++) {
      $this->buckets[] = -1;
    }
  }

  public
  function toString() {
    return sprintf("<Buckets %s>", $this->range->toString());
  }

  private
  function _assign($bucketRange, $expId) {
    $start = $bucketRange->getStart();
    if ($start < $this->range->getStart()) {
      $start = $this->range->getStart();
      Logger::error("experiment %d start %d out of range %s",
              $expId, $start, $bucketRange->toString());
    }
    $end = $bucketRange->getEnd();
    if ($end > $this->range->getEnd()) {
      $end = $this->range->getEnd();
      Logger::error("experiment %d end %d out of range %s",
              $expId, $end, $bucketRange->toString());
    }

    for ($i=$start - $this->range->getStart();
         $i<$end - $this->range->getStart(); $i++) {
      $this->buckets[$i] = $expId;
    }
    Logger::debug("assign exp %d to range %s", $expId, $bucketRange->toString());
  }

  public
  function assign($bucketRanges, $expId) {
    foreach ($bucketRanges as $bucketRange) {
      $this->_assign($bucketRange, $expId);
    }
  }

  public
  function locate($idx) {
    if ($this->range->contain($idx)) {
      $offset = $idx - $this->range->getStart();
      return $this->buckets[$offset];
    }
    return -1;
  }
}

class Domain {
  public
  function __construct($id, $bucketRange) {
    $this->id = intval($id);
    $this->range = $bucketRange;
  }

  public static
  function fromJson($json) {  // \exp_sys\Domain
    $range = $json->{'range'};
    return new self($json->{'id'}, BucketRange::fromJson($range));
  }

  public
  function getId() {
    return $this->id;
  }

  public
  function getBucketRange() {
    return $this->range;
  }

  public
  function toString() {
    return sprintf("<Domain %d, %s>", $this->id, $this->range->toString());
  }
}

class Layer {
  public
  function __construct($id, $domain) {
    $this->id = intval($id);
    $this->domain_id = $domain->getId();
    $this->buckets = new Buckets($domain->getBucketRange());

    $this->biased = false;
    srand(time());
  }

  public
  function toString() {
    return sprintf("<Layer %d, domain %d, %s>", $this->id,
                   $this->domain_id, $this->buckets->toString());
  }

  public
  function assign($exp) {
    Logger::debug("layer %d assign exp %d", $this->id, $exp->getId());
    $this->buckets->assign($exp->getRanges(), $exp->getId());
  }

  public
  function divert($idx) {
    return $this->buckets->locate($idx);
  }

  public
  function bias($biased=null) {
    if ($biased === null) {
      return $this->biased;
    } else {
      $this->biased = $biased?true:false;
    }
  }

  public
  function getId() {
    return $this->id;
  }
}

class Experiment {
  public
  function __construct($id, $layer_id, $start_time, $end_time,
          $diversion, $parameters, $conditions, $bucketRanges) {
    $this->id = intval($id);
    $this->layer_id = intval($layer_id);
    $this->start_time = $start_time;
    $this->end_time = $end_time;
    $this->diversion = $diversion;
    $this->parameters = $parameters;
    $this->conditions = $conditions;
    $this->ranges = $bucketRanges;
  }

  public
  function toString() {
    return sprintf("<Exp %d, layer %d(%s), time (%s~%s)>",
                   $this->id, $this->layer_id, $this->diversion,
                   $this->start_time, $this->end_time);
  }

  public static
  function fromJson($json) {
    $parameters = array();
    foreach ($json->{'parameters'} as $p) {
      $param = Parameter::fromJson($p);
      $parameters[$param->getName()] = $param->getValue();
    }
    $conditions = array();
    foreach ($json->{'conditions'} as $c) {
      $cond = Condition::fromJson($c);
      $conditions[$cond->getName()] = $cond->getArgs();
    }
    // TODO ranges seems useless
    $ranges = array();
    foreach ($json->{'ranges'} as $range) {
      $ranges[] = BucketRange::fromJson($range);
    }
    return new self($json->{'id'}, $json->{'layer_id'},
            $json->{'start_time'}, $json->{'end_time'}, $json->{'diversion'},  //
            $parameters, $conditions, $ranges);
  }

  public
  function validTime($timestamp) {
    $start = strtotime($this->start_time);
    $end   = strtotime($this->end_time);
    if ($timestamp >= $start && $timestamp <= $end) {
      return true;
    } else {
      return false;
    }
  }

  public
  function getRanges() {
    return $this->ranges;
  }

  public
  function getId() {
    return $this->id;
  }

  public
  function getLayerId() {
    return $this->layer_id;
  }

  public
  function getDiversion() {
    return $this->diversion;
  }

  public
  function getCond($name) {
    if (isset($this->conditions[$name])) {
      return $this->conditions[$name];
    }
    return null;
  }

  public
  function getCondMap() {
    return $this->conditions;
  }

  public
  function getParam($name) {
    if (isset($this->parameters[$name])) {
      return $this->parameters[$name];
    }
    return null;
  }

  public
  function getParamMap() {
    return $this->parameters;
  }
}

class Parameter {
  public
  function __construct($name, $value, $type=null) {
    $this->name = $name;
    $this->value = $value;
    $this->type = $type;
    // switch ($type) {
    //   case \exp_sys\Parameter\Type::BOOL:
    //     if (strtolower($value) == 'true') {
    //       $this->value = true;
    //     } else {
    //       $this->value = false;
    //     }
    //     break;
    //   case \exp_sys\Parameter\Type::INT:
    //     $this->value = intval($value);
    //     break;
    //   case \exp_sys\Parameter\Type::DOUBLE:
    //     $this->value = doubleval($value);
    //     break;
    //   default:
    //     $this->value = strval($value);
    //     break;
    // }
  }

  public static
  function fromJson($json) {
    return new self($json->{'name'}, $json->{'value'}, $json->{'type'});
  }

  public
  function getName() {
    return $this->name;
  }

  public
  function getValue() {
    return $this->value;
  }
}

class Condition {
  public
  function __construct($name, $args) {
    $this->name = $name;
    $this->args = $args;
  }

  public static
  function fromJson($json) {
    return new self($json->{'name'}, $json->{'args'});
  }

  public
  function getName() {
    return $this->name;
  }

  public
  function getArgs() {
    return $this->args;
  }
}